Add Prometheus total CPU metric and allow per-cpu stats to be disabled

bboreham opened the pull request:
Fixes #1513 

I'm a little uneasy that disabling the per-cpu stats means we use a different source for the data.
I could have it still go through the loop, but is that necessary?

cc @dashpole

dashpole commented:
I think we should collect the total usage the same way regardless of whether we are ignoring per-cpu stats.  If we are ignoring them, just skip that portion.

I would be interested to see how the CpuStats.Total compares with summing the per-cpu stats.

dashpole commented:
looking at the libcontainer implementation, it reads from `cpuacct.usage` to populate `stats.CpuStats.CpuUsage.TotalUsage`.  It reads from `cpuacct.usage_percpu` to populate `stats.CpuStats.CpuUsage.PercpuUsage`.  I think using CpuStats.Total to populate our total cpu usage makes more sense.  We should use that regardless of whether or not we are collecting per-cpu usage.

bboreham commented:
oops, just changed it the other way.  hold on...

dashpole commented:
Can you do a sanity check to see if it gives approximately correct values?  You can use the [stress container](https://github.com/vishh/stress), by using image `k8s.gcr.io/stress:v1` to test it, and run it with args: `--cpus=X`, and see if it uses the correct quantity of CPU.

bboreham commented:
Sure, will get to that later.

It appears there is a test which checks that CPU usage is computed by adding.

bboreham commented:
I have fixed the failing test, and changed my mind about the metric name.

I believe most people are currently using queries like

    sum (rate(container_cpu_usage_seconds_total{}[5m])) by (some_criteria)

If we change the name then everybody has to go round and edit all those queries.
Instead, if we keep the same name and just switch from `cpu01`, `cpu02`, etc., to `total` under the `cpu` label if `percpu` is disabled, then everything will Just Work.

Still haven't done the sanity check.

dashpole commented on metrics/prometheus.go:
```diff
@@ -150,10 +150,16 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc) *PrometheusCo
 			}, {
 				name:        "container_cpu_usage_seconds_total",
-				help:        "Cumulative cpu time consumed per cpu in seconds.",
+				help:        "Cumulative cpu time consumed in seconds.",
```
IIUC, users either get total OR per-cpu?  Why not just add a new metric for total?

bboreham replied:
See previous comment.

dashpole replied:
Ok, I agree with this.

dashpole commented:
Oh, and please also squash your commits.

dashpole commented:
@bboreham have you done your sanity check?  I would like to get this in the next release.

bboreham commented:
Thanks for the reminder - I got sidetracked onto 50 other things. Will squash and re-check tomorrow.

bboreham commented:
OK, I have run the `k8s.gcr.io/stress:v1` container a few times with different parameters, while running my build of cAdvisor from this branch, and checked the CPU metrics tallied with expectations.

I had forgotten to check `>0` before returning a CPU value so added that, for the percpu-disabled case.

And squashed commits as requested.
